### PR TITLE
Fix #12430: mvnup upgrade strategies and compatibility improvements

### PR DESCRIPTION
**Pull request by gnodet**
## Summary

Consolidates #12430 and #12443 into a single PR, plus new features:

### New strategies
- **`RepositoryHttpsUpgradeStrategy`** — converts HTTP repository URLs to HTTPS (blocked since Maven 3.8.1+), with well-known canonical URL mappings for Maven Central and Apache repositories
- **`EnforcerVersionRangeStrategy`** — widens `requireMavenVersion` ranges that exclude Maven 4 (e.g., `[3.8.8,4)` → `[3.8.8,5)`)
- **`ResourceFilteringStrategy`** (#12455) — adds `nonFilteredFileExtensions` to the `maven-resources-plugin` when resource filtering is enabled, preventing `MalformedInputException` on binary files (`.xlsx`, `.docx`, `.png`, etc.)
- **`DeduplicateDependenciesStrategy`** — standalone strategy for removing duplicate `<dependency>` and `<plugin>` declarations using last-wins semantics (extracted from `CompatibilityFixStrategy`)

### Quarkus plugin improvements
- **Skip upgrade for Quarkus 2.x** — detects the Quarkus platform version from `dependencyManagement` (quarkus-bom) or well-known properties (`quarkus.platform.version`, `quarkus.version`) and skips the `quarkus-maven-plugin` upgrade when the project uses Quarkus 2.x, preventing `NoSuchMethodError` build failures
- Quarkus plugin upgrade support, BOM version decoupling, and version gap warnings (from #12430)

### Bug fixes
- Fix mvnup comment formatting for override comments (from #12443)

### Other improvements (from #12430)
- Warn about property-interpolated module paths
- Warn about CI-friendly `${revision}` missing dependency versions
- Auto-upgrade `gmavenplus-plugin` to Maven 4-compatible version
- Upgrade DomTrip to 1.6.0 (from #12443)

### Removed (moved to #12464)
- ~~Silently ignore unrecognized CLI options from `.mvn/maven.config`~~ — root cause is that `bin/mvn` passes `$MAVEN_ARGS` to mvnup unconditionally; fix belongs in the shell script

Supersedes #12430 and #12443 (commits cherry-picked).

## Test plan
- [x] 27 tests for `RepositoryHttpsUpgradeStrategy`
- [x] 22 tests for `EnforcerVersionRangeStrategy`
- [x] 20 tests for `ResourceFilteringStrategy`
- [x] 15 tests for `DeduplicateDependenciesStrategy`
- [x] 11 tests for Quarkus platform version detection and 2.x skip behavior
- [x] Full module test suite passes (580 tests, 0 failures)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by gnodet — COMMENTED** [review]
Thorough review of this large PR (6900+ lines, 17 files) covering mvnup upgrade strategies and compatibility improvements.

**Verified findings (4 confirmed, 2 false positives dropped):**

1. **DeduplicateDependenciesStrategy — "last-wins" semantics documentation** (medium): The Javadoc on line 59 states "keeping the last declaration (matching Maven 3's last-wins behavior at runtime)", but Maven 3's dependency resolver actually uses **first-wins** for same-depth conflicts. For plugins, Maven 3 does merge later declarations over earlier ones (effectively last-wins for configuration), so last-wins is correct for plugins but not for dependencies. The old code in CompatibilityFixStrategy used first-wins for dependencies. Consider either updating the Javadoc to document this is an intentional behavior change, or switching to first-wins for dependencies.

2. **CommonsCliUpgradeOptions parse() retry loop** (low): When `CleanArgument.cleanArgs()` strips surrounding quotes, `UnrecognizedOptionException.getOption()` returns the post-cleaned option name while `currentArgs` has pre-cleaned strings. For a quoted arg like `"-ntp"`, `indexOf("-ntp")` would fail to match. The `idx < 0` guard correctly throws, so behavior degrades gracefully to an error rather than a silent bug. Very unlikely in practice.

3. **EnforcerVersionRangeStrategy regex scope** (low): The regex `4(?:\.\d+)*` matches any 4.x upper bound (4.1, 4.2, etc.), not just 4/4.0/4.0.0. The Javadoc examples only show the narrower case. Consider tightening to `4(?:\.0+)*` or updating the documentation.

4. **ResourceFilteringStrategy BINARY_EXTENSIONS** (low): The list of 23 entries is labeled "comprehensive" but omits common binary types (png, jpg/jpeg, gif, mp3, mp4, sqlite, p12/pfx). Consider expanding the list or softening the "comprehensive" claim.

5. **RepositoryHttpsUpgradeStrategy LinkedHashMap comment** (low): Comment claims ordering ensures "longer prefixes are checked before shorter ones", but entries are not actually ordered by length. Ordering doesn't matter for correctness since no key is a prefix of another, but the comment is misleading.

**False positives dropped by verifier:**
- Multi-segment version ranges: the regex DOES match union ranges correctly (lazy quantifier handles them)
- Missing test edge cases: existing test coverage is adequate

Overall: Well-structured PR with clean separation of strategies. No blocking issues — findings are informational comments and minor documentation improvements.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by gnodet** [comment]
Addressed the review feedback in cbdee27c39:

1. **DeduplicateDependenciesStrategy Javadoc** — Clarified that last-wins is an intentional choice (preserves the most recently added declaration), not a claim to match Maven 3's first-wins dependency resolution. Also noted that last-wins *does* match Maven 3's plugin configuration merging behavior.

2. **CommonsCliUpgradeOptions parse() retry loop** — Acknowledged; no change needed since behavior degrades gracefully to an error.

3. **EnforcerVersionRangeStrategy regex scope** — Updated the Javadoc to document that the regex intentionally matches any `4.x` upper bound (e.g., `4.1`, `4.2.1`) since all such bounds should be widened.

4. **ResourceFilteringStrategy BINARY_EXTENSIONS** — Added 10 missing extensions (`png`, `jpg`, `jpeg`, `gif`, `bmp`, `tiff`, `mp3`, `mp4`, `p12`, `pfx`) and softened the wording from "comprehensive" to "common" in both the class Javadoc and the field Javadoc.

5. **RepositoryHttpsUpgradeStrategy LinkedHashMap comment** — Fixed the misleading comment about prefix length ordering. Now correctly states the purpose is predictable insertion-order iteration.

**Review by gnodet — COMMENTED** [review]
## Re-review after commit cbdee27

The new commit addresses four of the five findings from the prior review:
- ✅ Class-level Javadoc corrected to clarify last-wins is intentional (not matching Maven 3 first-wins)
- ✅ EnforcerVersionRangeStrategy docs updated to document regex matching
- ✅ BINARY_EXTENSIONS expanded from 23 to 33 entries with softened wording
- ✅ LinkedHashMap comment fixed in RepositoryHttpsUpgradeStrategy

### Remaining issue

**Low — Javadoc inconsistency in `DeduplicateDependenciesStrategy`**

The class-level Javadoc was correctly updated to say:
> "Maven 3's dependency resolver uses first-wins for same-depth conflicts; last-wins is chosen here intentionally"

But the method-level Javadoc for `fixDuplicateDependenciesInSection` (~line 183) still says:
> "matching Maven 3's runtime behavior"

These contradict each other. Suggest changing the method Javadoc to just say "last-wins" without referencing Maven 3 behavior.

---

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/CompatibilityFixStrategy.java [hunk 1]
```diff
@@ -81,6 +85,15 @@ public class CompatibilityFixStrategy extends AbstractUpgradeStrategy {
...
     private static final Set<String> VALID_COMBINE_CHILDREN_VALUES = Set.of(COMBINE_APPEND, COMBINE_MERGE);
 
+    /**
+     * Known incompatible plugins where even the latest version fails with Maven 4.
+     * Maps plugin key (groupId:artifactId) to a description of the incompatibility.
+     * <p>
+     * Note: plugins that have a fixed version available should be added to
+     * {@link PluginUpgradeStrategy} instead, so mvnup can auto-upgrade them.
+     */
+    private static final Map<String, String> KNOWN_INCOMPATIBLE_PLUGINS = Map.of();
```
Comment (elharo) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/CompatibilityFixStrategy.java:95]: I appreciate special casing known incompatibilities. I wish we did more of this.

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/CompatibilityFixStrategy.java [hunk 2]
```diff
@@ -816,4 +662,148 @@ private Path findParentPomInMap(
                 .map(Map.Entry::getKey)
                 .orElse(null);
     }
+
+    // ---- Warning-only checks for Maven 4 compatibility issues that cannot be auto-fixed ----
+
+    /**
+     * Warns about plugins known to be incompatible with Maven 4 even at their latest version.
+     * These plugins call methods on immutable API objects or use removed internal APIs
```
Comment (elharo) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/CompatibilityFixStrategy.java:670]: Arte these the only cases? Since 4.x is a major upgrade it's not a surprise that the API has changed incompatibly in various ways. Maybe just say something like "These plugins do not yet work with Maven 4." or "There is not a version of this plugin that works with Maven 4 at the time of writing."

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/DeduplicateDependenciesStrategy.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+import static eu.maveniverse.domtrip.maven.MavenPomElements.Plugins.MAVEN_PLUGIN_PREFIX;
+
+/**
+ * Strategy for deduplicating dependency and plugin declarations in POM files.
+ *
+ * <p>Maven 4 rejects duplicate {@code <dependency>} entries (same groupId, artifactId,
+ * type, and classifier) that Maven 3 silently accepted. This strategy scans each POM's
+ * {@code <dependencies>} and {@code <dependencyManagement>/<dependencies>} sections and
+ * removes duplicates using last-wins semantics (keeping the last declaration). Note that
+ * Maven 3's dependency resolver uses first-wins for same-depth conflicts; last-wins is
```
Comment (elharo) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnup/goals/DeduplicateDependenciesStrategy.java:58]: I wouldn't assume that at all. When I've added a duplicate dependency (which I've done more than a few times) it's landed almost anywhere in the list depending mostly on where and what I was copying and pasting into. There is no expectation that dependencies are added only at the end. I think it's better to keep the first to more closely reflect the current dependency tree.

File: pom.xml [hunk 4]
```diff
@@ -146,7 +146,7 @@ under the License.
     <byteBuddyVersion>1.18.11</byteBuddyVersion>
     <classWorldsVersion>2.12.0</classWorldsVersion>
     <commonsCliVersion>1.11.0</commonsCliVersion>
-    <domtripVersion>1.5.2</domtripVersion>
+    <domtripVersion>1.6.0</domtripVersion>
```
Comment (elharo) [inline pom.xml:149]: This should likely be a separate PR

**Review by cstamas — APPROVED** [review]
LGTM

**Review by gnodet — COMMENTED** [review]
## Re-review — New Commit (Revert)

The new commit (93b682ab) cleanly reverts the unrecognized-option retry loop from `CommonsCliUpgradeOptions.CLIManager.parse()`, which was introduced earlier in this PR (Fix #12431). The revert removes both the implementation and its 6 test methods.

This is the right call — the retry loop silently swallowed unknown options, masking real user errors. The root cause (`mvn` launcher unconditionally passing `$MAVEN_ARGS` to mvnup) will be fixed at the shell-script level in #12464 instead.

The revert is complete and correct:
- `parse()` override and all associated imports removed
- All 6 test methods for the retry loop removed from `PluginUpgradeCliTest`
- No orphaned references remain
- Commit message clearly cross-references #12464

No new issues found. ✅

---

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by cstamas — APPROVED** [review]
